model/pod/dispatcher/db.go

```diff
@@ -0,0 +1,148 @@
+	return db.Update(Collection, query, update)
+}
+
+// UpsertOne updates an existing pod dispatcher if it exists based on the
+// query; otherwise, it inserts a new pod dispatcher.
+func UpsertOne(query, update interface{}) (*adb.ChangeInfo, error) {
+	return db.Upsert(Collection, query, update)
+}
+
+// FindOneByID finds one pod dispatcher by its ID.
+func FindOneByID(id string) (*PodDispatcher, error) {
+	return FindOne(db.Query(bson.M{
+		IDKey: id,
+	}))
+}
+
+// ByGroupID returns the query to find a pod dispatcher by its group ID.
+func ByGroupID(groupID string) bson.M {
+	return bson.M{
+		GroupIDKey: groupID,
+	}
+}
+
+// FindOneByGroupID finds one pod dispatcher by its group ID.
+func FindOneByGroupID(groupID string) (*PodDispatcher, error) {
+	return FindOne(db.Query(ByGroupID(groupID)))
+}
+
+// Allocate sets up the given intent pod to the given task for dispatching.
+func Allocate(ctx context.Context, env evergreen.Environment, t *task.Task, p *pod.Pod) (*PodDispatcher, error) {
+	mongoClient := env.Client()
+	session, err := mongoClient.StartSession()
+	if err != nil {
+		return nil, errors.Wrap(err, "starting transaction session")
+	}
+	defer session.EndSession(ctx)
+
+	pd := &PodDispatcher{}
+	allocateDispatcher := func(sessCtx mongo.SessionContext) (interface{}, error) {
+		groupID := GetGroupID(t)
+		if err := env.DB().Collection(Collection).FindOne(sessCtx, ByGroupID(groupID)).Decode(pd); err != nil && !adb.ResultsNotFound(err) {
+			return nil, errors.Wrap(err, "checking for existing pod dispatcher")
+		} else if adb.ResultsNotFound(err) {
+			newDispatcher := NewPodDispatcher(groupID, []string{t.Id}, []string{p.ID})
+			pd = &newDispatcher
+		} else {
+			pd.PodIDs = append(pd.PodIDs, p.ID)
+
+			if !utility.StringSliceContains(pd.TaskIDs, t.Id) {
+				pd.TaskIDs = append(pd.TaskIDs, t.Id)
+			}
+		}
+
+		if _, err := env.DB().Collection(Collection).UpdateOne(sessCtx, pd.atomicUpsertQuery(), pd.atomicUpsertUpdate(), options.Update().SetUpsert(true)); err != nil {
+			return nil, errors.Wrap(err, "upserting pod dispatcher")
+		}
+		pd.ModificationCount++
+
+		if _, err := env.DB().Collection(pod.Collection).InsertOne(sessCtx, p); err != nil {
+			return nil, errors.Wrap(err, "inserting new intent pod")
+		}
+
+		// Only allow the task to transition state if it's currently in a state
+		// where it needs a pod to be allocated.
+		update, err := env.DB().Collection(task.Collection).UpdateOne(sessCtx, bson.M{
+			task.IdKey:        t.Id,
+			task.StatusKey:    evergreen.TaskContainerUnallocated,
+			task.ActivatedKey: true,
+			task.PriorityKey:  bson.M{"$gt": evergreen.DisabledTaskPriority},
+		}, bson.M{
+			"$set": bson.M{
+				task.StatusKey: evergreen.TaskContainerAllocated,
+			},
+		})
+		if err != nil {
+			return nil, errors.Wrap(err, "marking task as container allocated")
+		}
+		if update.ModifiedCount == 0 {
+			return nil, errors.New("task status was not updated")
+		}
+		t.Status = evergreen.TaskContainerAllocated
```

Comment:
Is it ever possible that `t` will have `Activated` set to `false` and `PriorityKey` set to `evergreen.DisabledTaskPriority` in-mem but not in the DB? In which case, we should probably update those fields here as well.

Reply:
They don't need to be updated because they're part of the query, not the update. In order for the transaction to succeed, the task has to have both of those fields be the same as they were when the task was first loaded from the DB to prove that the in-memory task state hasn't become stale.